tests/rc_string_ut/rc_string_ut.c

```diff
@@ -26,40 +42,432 @@ static void on_umock_c_error(UMOCK_C_ERROR_CODE error_code)
     ASSERT_FAIL("umock_c reported error :%" PRI_MU_ENUM "", MU_ENUM_VALUE(UMOCK_C_ERROR_CODE, error_code));
 }
 
+// this function is used for tests just to make sure we do not confuse it with regular malloc
+MOCK_FUNCTION_WITH_CODE(, void*, test_malloc_func, size_t, size_needed)
+MOCK_FUNCTION_END((unsigned char*)my_gballoc_malloc(size_needed + 1) + 1)
+
+MOCK_FUNCTION_WITH_CODE(, void, test_free_func, void*, context)
+    my_gballoc_free((unsigned char*)context - 1);
+MOCK_FUNCTION_END()
+
+MOCK_FUNCTION_WITH_CODE(, void, test_free_func_do_nothing, void*, context)
+MOCK_FUNCTION_END()
+
 BEGIN_TEST_SUITE(rc_string_unittests)
 
-    TEST_SUITE_INITIALIZE(suite_initialize)
-    {
-        g_testByTest = TEST_MUTEX_CREATE();
-        ASSERT_IS_NOT_NULL(g_testByTest);
+TEST_SUITE_INITIALIZE(suite_initialize)
+{
+    g_testByTest = TEST_MUTEX_CREATE();
+    ASSERT_IS_NOT_NULL(g_testByTest);
+
+    ASSERT_ARE_EQUAL(int, 0, umock_c_init(on_umock_c_error));
+    ASSERT_ARE_EQUAL(int, 0, umocktypes_charptr_register_types());
+
+    REGISTER_GLOBAL_MOCK_HOOK(gballoc_malloc, my_gballoc_malloc);
+    REGISTER_GLOBAL_MOCK_HOOK(gballoc_free, my_gballoc_free);
+
+    REGISTER_GLOBAL_MOCK_FAIL_RETURN(gballoc_malloc, NULL);
+}
+
+TEST_SUITE_CLEANUP(suite_cleanup)
+{
+    umock_c_deinit();
 
-        ASSERT_ARE_EQUAL(int, 0, umock_c_init(on_umock_c_error));
-        ASSERT_ARE_EQUAL(int, 0, umocktypes_charptr_register_types());
+    TEST_MUTEX_DESTROY(g_testByTest);
+}
+
+TEST_FUNCTION_INITIALIZE(function_init)
+{
+    if (TEST_MUTEX_ACQUIRE(g_testByTest))
+    {
+        ASSERT_FAIL("our mutex is ABANDONED. Failure in test framework");
     }
 
-    TEST_SUITE_CLEANUP(suite_cleanup)
+    ASSERT_ARE_EQUAL(int, 0, umock_c_negative_tests_init());
+
+    umock_c_reset_all_calls();
+}
+
+TEST_FUNCTION_CLEANUP(function_cleanup)
+{
+    umock_c_negative_tests_deinit();
+    TEST_MUTEX_RELEASE(g_testByTest);
+}
+
+/* rc_string_create */
+
+/*Tests_SRS_RC_STRING_01_001: [ If string is NULL, rc_string_create shall fail and return NULL. ]*/
+TEST_FUNCTION(rc_string_create_with_NULL_fails)
+{
+    // arrange
+
+    // act
+    THANDLE(RC_STRING) rc_string = rc_string_create(NULL);
+
+    // assert
+    ASSERT_ARE_EQUAL(char_ptr, umock_c_get_expected_calls(), umock_c_get_actual_calls());
+    ASSERT_IS_NULL(rc_string);
+}
+
+/* Tests_SRS_RC_STRING_01_002: [ Otherwise, rc_string_create shall determine the length of string. ]*/
+/* Tests_SRS_RC_STRING_01_003: [ rc_string_create shall allocate memory for the THANDLE(RC_STRING), ensuring all the bytes in string can be copied (including the zero terminator). ]*/
+/* Tests_SRS_RC_STRING_01_004: [ rc_string_create shall copy the string memory (including the NULL terminator). ]*/
+/* Tests_SRS_RC_STRING_01_005: [ rc_string_create shall succeed and return a non-NULL handle. ]*/
+TEST_FUNCTION(rc_string_create_succeeds)
+{
+    // arrange
+    STRICT_EXPECTED_CALL(malloc(IGNORED_ARG));
```

Comment:
>STRICT_EXPECTED_CALL [](start = 4, length = 20)

hm.... either here is missing a interlocked_exchange... OR thandle didn't get it yet... OR a merge was not done.

I would assume THANDLE did not get to use juicy interlocked_xxx APIs. It is fine.

Reply:
Yeah, I think it did not get it ... yet.

---
In reply to: [460260312](https://github.com/Azure/azure-c-util/pull/52#discussion_r460260312) [](ancestors = 460260312)